### PR TITLE
Fix dirty collection attribute refresh

### PR DESCRIPTION
**Pull request by VenelinMartinov**
This fixes an issue with empty collection attributes refreshing dirty under PlanResourceChange. Note that this is an issue present in TF as well.
It comes from the fact that some (most?) providers don't distinguish between null and empty collection values, for example both AWS and GCP labels have this issue.

We've opted for a targeted fix in Refresh - this maintains our TF-conformance on Create/Update.
The fix is that on refresh we go over the read values and compare with the previous state. Any properties which:
1. Are a TF attribute
2. Are a collection
3. Are null or empty
4. Are different to the previous state

get "normalized" to what was in the state. This avoids us displaying excessive dirty refreshes for these values.

Note that the actual fix is from https://github.com/pulumi/pulumi-terraform-bridge/pull/2073 and due to @t0yv0
A lot of the commit history here is from https://github.com/pulumi/pulumi-terraform-bridge/pull/2072 which separated the tests from here to facilitate easier reviews. I should have merged here before deleting 🤷 

Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/2047
Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1967

### CODE REVIEW

File: pkg/tfgen/convert_cli_test.go [hunk 1]
```diff
@@ -57,6 +57,7 @@ func TestConvertViaPulumiCLI(t *testing.T) {
 		t.Skipf("Skipping on Windows due to a test setup issue")
 	}
 	t.Setenv("PULUMI_CONVERT", "1")
+	t.Setenv("DISABLE_AUTOMATIC_PLUGIN_ACQUISITION", "true")
```
Comment (VenelinMartinov) [inline pkg/tfgen/convert_cli_test.go:60]: fixes a flaky test, I'll merge this separately

**Comment by VenelinMartinov** [comment]
I've ran downstream tests here and they are good: https://github.com/pulumi/pulumi-terraform-bridge/issues/1967#issuecomment-2161096367

**Comment by VenelinMartinov** [comment]
Note https://github.com/pulumi/pulumi/pull/16146 is relevant here - it might fix some of the dirty refreshes - should we wait for that and see if we still need to ship this?

File: pkg/tfshim/sdk-v2/walk.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
+package sdkv2
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/walk"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+func findSchemaContext(topLevel *schema.Resource, path walk.SchemaPath) schemaContext {
```
Comment (t0yv0) [inline pkg/tfshim/sdk-v2/walk.go:9]: @iwahbe appreciate quick eyes on this; this maybe something that can be expressed more fluently or lifted to the shim.* layer. What I find handy is working around the inability to reference internal types directly with something like this:

```
type attrSchemaContext struct {
	resource *schema.Resource
	name     string
}
```

Which now gives access to this internal machinery via `ctx.resource.CoreConfigSchema().Attributes[a.name].Optional` etc.

**Review by t0yv0 — APPROVED** [review]
Thank you for taking over and bringing to completion @VenelinMartinov ! Still appreciate thorough review to see if any potential panics are missed etc.

**Comment by VenelinMartinov** [comment]
Found a regression caused by this change: https://github.com/pulumi/pulumi-gcp/issues/2079#issuecomment-2163296305

Without the change the `Firewall` resource refreshes dirty: `gcp:compute:Firewall ts-firewall updated (0.48s) [diff: +sourceRanges,sourceServiceAccounts,targetServiceAccounts,targetTags]`

But that's what happens in TF too:

```
  # google_compute_firewall.py_firewall has changed
  ~ resource "google_compute_firewall" "py_firewall" {
        id                      = "projects/pulumi-development/global/firewalls/py-firewall"
        name                    = "py-firewall"
      + source_ranges           = []
      + source_service_accounts = []
      + target_service_accounts = []
      + target_tags             = []
        # (10 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }
```

With the change it still refreshes dirty but in a different parameter.

**Comment by t0yv0** [comment]
The change intentionally deviates from TF behavior so that by itself is not a problem. 

> With the change it still refreshes dirty but in a different parameter.

Curious any more details here?

**Comment by VenelinMartinov** [comment]
https://github.com/pulumi/pulumi-gcp/issues/2079#issuecomment-2163296305 has the Read GRPC call. It's in the `allows` property, the member with "icmp" - ports changes from [] to nil in the Read

**Comment by t0yv0** [comment]
I'd like to debug the GCP example. Setting this up shortly.

**Comment by VenelinMartinov** [comment]
Ah, I ran it with --refresh:

```
~ allows               : [
          ~ [0]: {
                  ~ ports   : [
                      + [0]: "22"
                      + [1]: "80"
                    ]
                  ~ protocol: "icmp" => "tcp"
                }
          ~ [1]: {
                  - ports   : [
                  -     [0]: "22"
                  -     [1]: "80"
                    ]
                  ~ protocol: "tcp" => "icmp"
                }
        ]
```

This is likely a reordering issue, not a problem with the values.

**Comment by t0yv0** [comment]
I've had a look and I know what's wrong with the GCP example. 

The correction kicks in with this:

```
  REPLACING allow.$.ports FROM cty.ListValEmpty(cty.String) TO cty.NullVal(cty.List(cty.String))
```

But what's unexpected here is that the state value is actually cty.ListValEmpty(cty.String) so this should not apply.

How does this happen? Turns out "github.com/hashicorp/go-cty/cty" is buggy around sets:

cty.Path apply expects only ints or strings but cty.Transform injects entire set element values into IndexStep. Min repro:

```
func TestSimpleSet(t *testing.T) {
	vtop := cty.SetVal([]cty.Value{
		cty.ObjectVal(map[string]cty.Value{
			"foo": cty.StringVal("bar"),
		}),
	})

	cty.Transform(vtop, func(p cty.Path, v cty.Value) (cty.Value, error) {
		vv, err := p.Apply(vtop)
		if err != nil || vv.Equals(v).False() {
			t.Logf("err %v", err)
			t.Logf("path %v", p)
		}
		return v, nil
	})
}
```

As a result, looking for a matching **state** element under the path fails to find one, and assumes it is missing or explicit null. 

This is not good enough:

```
	matchingOldStateValue := func(t cty.Type, p cty.Path) cty.Value {
		if v, err := p.Apply(oldState); err == nil {
			return v
		}
		return cty.NullVal(t)
	}
```

**Comment by t0yv0** [comment]
https://github.com/zclconf/go-cty/issues/180 is where it's bottoming out. As a workaround we can skip paths that contain set elements and just bail for this transformation. We can bail on error from p.Apply. Or we can rewrite p.Apply into a "better" version.

File: pkg/tfshim/sdk-v2/walk.go (hunk 2 above)
Comment (iwahbe) [reply pkg/tfshim/sdk-v2/walk.go:9]: I'm not sure we want to expose nesting level info to `shim.*`. Nothing jumps out to me.

File: pkg/tfshim/sdk-v2/walk.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+func (b *blockSchemaContext) attribute(name string) schemaContext {
+	s := b.resource.CoreConfigSchema()
+	_, isAttr := s.Attributes[name]
+	if isAttr {
+		return &attrSchemaContext{b.resource, name}
+	}
+	blk, isBlock := s.BlockTypes[name]
+	if isBlock {
+		switch int(blk.Nesting) {
+		case 1:
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply pkg/tfshim/sdk-v2/walk.go]: The model is internal and not accessible, but looks like we have a copy of the model we can use.

**Comment by t0yv0** [comment]
We're weighing now whether to include this change or not. This introduces an intentional deviation in Read results for Pulumi bridged providers vs TF baseline to compensate for old normalizeNullValues code in TF that intentionally confuses empty and null collections leading to dirty refresh in Pulumi. With this change the Read result is modified to match the state and avoid having a diff. Under the new CLI https://github.com/pulumi/pulumi/releases/tag/v3.120.0 some (or all?) these dirty refresh scenarios may refresh cleanly.

Should we still include?

**Comment by t0yv0** [comment]
Reasons to include: the change is fairly surgical and well motivated, hedge against any subsequent changes in the CLI

Reasons not to include: any change carries some unforeseen risk, having an intentional difference from TF behavior creates more maintenance

Information that could help decide: 

- can we confirm that existing test cases pass on the latest CLI without the change?
- are there any more refresh changes to try out?  (@t0yv0 can scan through refresh tickets in AWS classic to try them)

**Comment by t0yv0** [comment]
@mjeffryes sounds like you had some input here.

**Comment by mjeffryes** [comment]
Sure. 
- https://github.com/pulumi/pulumi/pull/16146 has landed and appears to address all our test failure for "dirty refresh" and should mitigate that problem for users. It does change the outputs in state, so that means that resources that depend on that output *might* try to update after a "clean" refresh (assuming the null/empty distinction matters to the dependent resource).
- The normalization change we've prepared would completely hide when upstream returns a nil instead of empty or vice-versa, so it would suppress updates to dependent resources even if the difference between null and empty was somehow significant.

- It's honestly not clear how often this difference in behavior is actually significant, but the behavior without normalization is both more transparent to the user and more consistent with the upstream TF provider. Hence, I'm inclined to try rolling out PRC without normalization and see if users encounter any real situations where they wish we would have normalized to suppress the update.

**Comment by VenelinMartinov** [comment]
Running RPC downstream tests without this PR here: https://github.com/pulumi/pulumi-terraform-bridge/issues/1962#issuecomment-2167990221

**Comment by VenelinMartinov** [comment]
Also note that under the new refresh behaviour the tests fixed in this PR no longer fail without it: https://github.com/pulumi/pulumi-terraform-bridge/pull/2100

**Comment by t0yv0** [comment]
Yes sounds like these things got fixed in the CLI so perhaps this PR is not necessary anymore. Is that too simplistic of a take?

**Pull request closed without merging**